In [1]:
import pandas as pd
import numpy as np

In [132]:
data_forct = pd.read_excel('C:\\Users\\Ramshankar\\OneDrive - iLink Systems Inc\\Documents\\Costco\\Code\\analysis\\Forecasted_invntry_stock_Anlys_6months.xlsx', sheet_name= 'Sheet1')

In [133]:
data_forct

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O        84.64   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O        65.47   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O        81.99   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O        82.37   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O        72.33   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        55.94   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        81.92   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        38.68   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        76.53   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        68.86   

     On Order Quantity  In Transit Quantity  NSI Quantity  Quantity  Amount  \
0                99.41                37.29             0     55.80  334.24   
1                71.47                11.99             0     41.94  251.21   
2                92.42                44.47             0     65.50  392.34   
3                98.76                42.11             0     27.73  166.11   
4                82.90                 4.71             0     19.68  117.88   
..                 ...                  ...           ...       ...     ...   
429              45.11                45.54             0    113.51  146.43   
430              47.40                68.25             0     97.40  125.65   
431              26.79                20.34             0     93.24  120.28   
432              76.63                64.16             0     94.38  121.75   
433              66.40                52.27             0     88.66  114.37   

     Forecasted_Qty  Stock_Qty     Stocks  Stock_flag  
0             46.71      24.00     Onhand           3  
1             54.25      48.00  Overstock           1  
2             49.95      16.00     Onhand           3  
3             38.10      12.00     Onhand           3  
4             43.47      49.00  Overstock           1  
..              ...        ...        ...         ...  
429          115.52      59.00  Overstock           1  
430          118.61      15.84     Onhand           3  
431          132.86     101.00  Overstock           1  
432           87.94      24.00     Onhand           3  
433           82.54      28.00     Onhand           3  

[434 rows x 13 columns]

In [35]:
data1= data_forct[['Date', 'Location', 'Item', 'Forecasted_Qty', 'Stock_Qty', 'Stocks']]

In [39]:
data1= data1.rename(columns= {'Stock_Qty': 'Stock_Qty1', 'Stocks': 'Stocks1'})

In [40]:
data1

Date                 Location               Item  Forecasted_Qty  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O           46.71   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O           54.25   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O           49.95   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O           38.10   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O           43.47   
..         ...                      ...                ...             ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          115.52   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          118.61   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          132.86   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           87.94   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           82.54   

     Stock_Qty1    Stocks1  
0         24.00     Onhand  
1         48.00  Overstock  
2         16.00     Onhand  
3         12.00     Onhand  
4         49.00  Overstock  
..          ...        ...  
429       59.00  Overstock  
430       15.84     Onhand  
431      101.00  Overstock  
432       24.00     Onhand  
433       28.00     Onhand  

[434 rows x 6 columns]

In [54]:
data_inv = pd.read_excel('C:\\Users\\Ramshankar\\OneDrive - iLink Systems Inc\\Documents\\Costco\\Code\\analysis\\sales_invntry_Actual_stock_6months_weekly.xlsx', sheet_name= 'Sheet1')

In [55]:
data_inv

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O           85   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O           65   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O           82   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O           82   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O           72   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           56   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           42   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           39   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           77   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           69   

     On Order Quantity  In Transit Quantity  NSI Quantity  Quantity  Amount  \
0                   99                   37             0     55.80  334.24   
1                   71                   12             0     41.94  251.21   
2                   92                   44             0     65.50  392.34   
3                   99                   42             0     27.73  166.11   
4                   83                    5             0     19.68  117.88   
..                 ...                  ...           ...       ...     ...   
429                 45                   46             0    113.51  146.43   
430                 47                   38             0     97.40  125.65   
431                 27                   20             0     93.24  120.28   
432                 77                   64             0     94.38  121.75   
433                 66                   52             0     88.66  114.37   

     Forecasted_Qty  Stock_Qty     Stocks  Stock_flag  
0             46.71         33  Overstock           1  
1             54.25         36  Overstock           1  
2             49.95         32  Overstock           1  
3             38.10          3     Onhand           3  
4             43.47         26  Overstock           1  
..              ...        ...        ...         ...  
429          115.52         57  Overstock           1  
430          118.61         64  Overstock           1  
431          132.86         61  Overstock           1  
432           87.94         30  Overstock           1  
433           82.54         34  Overstock           1  

[434 rows x 13 columns]

In [56]:
data2= data_inv[['Date', 'Location', 'Item', 'Forecasted_Qty', 'Stock_Qty', 'Stocks']]

In [57]:
data2= data2.rename(columns= {'Forecasted_Qty': 'Forecasted_Qty2', 'Stock_Qty': 'Stock_Qty2', 'Stocks': 'Stocks2'})

In [58]:
data2

Date                 Location               Item  Forecasted_Qty2  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O            46.71   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O            54.25   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O            49.95   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O            38.10   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O            43.47   
..         ...                      ...                ...              ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           115.52   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           118.61   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           132.86   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN            87.94   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN            82.54   

     Stock_Qty2    Stocks2  
0            33  Overstock  
1            36  Overstock  
2            32  Overstock  
3             3     Onhand  
4            26  Overstock  
..          ...        ...  
429          57  Overstock  
430          64  Overstock  
431          61  Overstock  
432          30  Overstock  
433          34  Overstock  

[434 rows x 6 columns]

In [84]:
df= pd.merge(data1, data2, on=['Date', 'Location', 'Item'], how='inner')

In [85]:
df

Date                 Location               Item  Forecasted_Qty  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O           46.71   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O           54.25   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O           49.95   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O           38.10   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O           43.47   
..         ...                      ...                ...             ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          115.52   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          118.61   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          132.86   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           87.94   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           82.54   

     Stock_Qty1    Stocks1  Forecasted_Qty2  Stock_Qty2    Stocks2  
0         24.00     Onhand            46.71          33  Overstock  
1         48.00  Overstock            54.25          36  Overstock  
2         16.00     Onhand            49.95          32  Overstock  
3         12.00     Onhand            38.10           3     Onhand  
4         49.00  Overstock            43.47          26  Overstock  
..          ...        ...              ...         ...        ...  
429       59.00  Overstock           115.52          57  Overstock  
430       15.84     Onhand           118.61          64  Overstock  
431      101.00  Overstock           132.86          61  Overstock  
432       24.00     Onhand            87.94          30  Overstock  
433       28.00     Onhand            82.54          34  Overstock  

[434 rows x 9 columns]

In [79]:
df['diff']= ((abs(df['Stock_Qty1']-df['Stock_Qty2']))/((df['Stock_Qty1']+df['Stock_Qty2'])/2)*100)

In [ ]:
df['diff']= (abs(df['Stock_Qty1']-df['Stock_Qty2'])

In [121]:
def custom_calc(x):
    return ((abs(x[0] - x[1]))/ x[1]*100).round(2)


In [122]:
df['diff']= df[['Stock_Qty1', 'Stock_Qty2']].apply(custom_calc, axis=1)

C:\Users\Ramshankar\AppData\Local\Temp\ipykernel_5784\248686723.py:2: RuntimeWarning: divide by zero encountered in double_scalars
  return ((abs(x[0] - x[1]))/ x[1]*100).round(2)


In [123]:
df

Date                 Location               Item  Forecasted_Qty  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O           46.71   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O           54.25   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O           49.95   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O           38.10   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O           43.47   
..         ...                      ...                ...             ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          115.52   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          118.61   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          132.86   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           87.94   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           82.54   

     Stock_Qty1    Stocks1  Forecasted_Qty2  Stock_Qty2    Stocks2    diff  
0         24.00     Onhand            46.71          33  Overstock   27.27  
1         48.00  Overstock            54.25          36  Overstock   33.33  
2         16.00     Onhand            49.95          32  Overstock   50.00  
3         12.00     Onhand            38.10           3     Onhand  300.00  
4         49.00  Overstock            43.47          26  Overstock   88.46  
..          ...        ...              ...         ...        ...     ...  
429       59.00  Overstock           115.52          57  Overstock    3.51  
430       15.84     Onhand           118.61          64  Overstock   75.25  
431      101.00  Overstock           132.86          61  Overstock   65.57  
432       24.00     Onhand            87.94          30  Overstock   20.00  
433       28.00     Onhand            82.54          34  Overstock   17.65  

[434 rows x 10 columns]

In [124]:
# def custom_calc(x):
#     return (((x[1] - x[0]) / x[0]) * 100).round(2)

# df['diff'] = df[['Stock_Qty1', 'Stock_Qty2']].apply(custom_calc, axis=1)

In [125]:
df

Date                 Location               Item  Forecasted_Qty  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O           46.71   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O           54.25   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O           49.95   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O           38.10   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O           43.47   
..         ...                      ...                ...             ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          115.52   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          118.61   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          132.86   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           87.94   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN           82.54   

     Stock_Qty1    Stocks1  Forecasted_Qty2  Stock_Qty2    Stocks2    diff  
0         24.00     Onhand            46.71          33  Overstock   27.27  
1         48.00  Overstock            54.25          36  Overstock   33.33  
2         16.00     Onhand            49.95          32  Overstock   50.00  
3         12.00     Onhand            38.10           3     Onhand  300.00  
4         49.00  Overstock            43.47          26  Overstock   88.46  
..          ...        ...              ...         ...        ...     ...  
429       59.00  Overstock           115.52          57  Overstock    3.51  
430       15.84     Onhand           118.61          64  Overstock   75.25  
431      101.00  Overstock           132.86          61  Overstock   65.57  
432       24.00     Onhand            87.94          30  Overstock   20.00  
433       28.00     Onhand            82.54          34  Overstock   17.65  

[434 rows x 10 columns]

In [126]:
df[df['Date']== '2022-07-10']

Date                 Location               Item  Forecasted_Qty  \
10  2022-07-10                  FREMONT  CHICKEN XXXXX - O           29.85   
41  2022-07-10                  FREMONT    DRUMSTICKS NNNN          694.35   
72  2022-07-10                  FREMONT        FILLET QQQQ           43.89   
103 2022-07-10                  HAYWARD  CHICKEN XXXXX - O           41.18   
134 2022-07-10                  HAYWARD    DRUMSTICKS NNNN          595.95   
165 2022-07-10                  HAYWARD        FILLET QQQQ           60.54   
196 2022-07-10                 LAKEWOOD  CHICKEN XXXXX - O           40.00   
227 2022-07-10                 LAKEWOOD    DRUMSTICKS NNNN          331.70   
258 2022-07-10                 LAKEWOOD        FILLET QQQQ           57.65   
289 2022-07-10              RICHMOND CA  CHICKEN XXXXX - O           62.45   
320 2022-07-10              RICHMOND CA    DRUMSTICKS NNNN          816.66   
351 2022-07-10              RICHMOND CA        FILLET QQQQ           54.66   
382 2022-07-10  S SAN FRANCISCO BUS CTR     BEEF OOOO - GT          366.94   
413 2022-07-10  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN          144.78   

     Stock_Qty1     Stocks1  Forecasted_Qty2  Stock_Qty2     Stocks2    diff  
10        -2.00  Understock            29.85           1      Onhand  300.00  
41       133.00   Overstock           694.35         209   Overstock   36.36  
72        29.00      Onhand            43.89          35   Overstock   17.14  
103        4.00      Onhand            41.18           7      Onhand   42.86  
134      130.00   Overstock           595.95         187   Overstock   30.48  
165       54.00   Overstock            60.54          57   Overstock    5.26  
196       68.00   Overstock            40.00          71   Overstock    4.23  
227      111.00   Overstock           331.70          33   Overstock  236.36  
258       56.00   Overstock            57.65          61   Overstock    8.20  
289      -24.00  Understock            62.45           7      Onhand  442.86  
320     -143.00  Understock           816.66        -126  Understock  -13.49  
351       52.00   Overstock            54.66          50   Overstock    4.00  
382       23.74      Onhand           366.94        -277  Understock -108.57  
413       22.25      Onhand           144.78         115   Overstock   80.65

In [128]:
df[df['Stocks1']== 'Onhand'].min()

Date               2022-05-01 00:00:00
Location                       FREMONT
Item                    BEEF OOOO - GT
Forecasted_Qty                   19.51
Stock_Qty1                        3.28
Stocks1                         Onhand
Forecasted_Qty2                  19.51
Stock_Qty2                        -357
Stocks2                         Onhand
diff                           -366.67
dtype: object

In [ ]:
Onhand
max_change=  400
min= -2520.0

Overstock
max= 421.276596
min= -1440.0

Understock:
max= 494.275274
min=  -1200.0



In [83]:
df[df['diff']== -600.0]

Date  Location               Item  Forecasted_Qty  Stock_Qty1  \
10  2022-07-10   FREMONT  CHICKEN XXXXX - O           29.85        -2.0   
247 2022-11-27  LAKEWOOD    DRUMSTICKS NNNN          190.71        -6.0   

        Stocks1  Forecasted_Qty2  Stock_Qty2 Stocks2   diff  
10   Understock            29.85           1  Onhand -600.0  
247  Understock           190.71           3  Onhand -600.0

Onhand
max_change=  600
min= -2520.0

Overstock
max= 421.276596
min= -1440.0

Understock:
max= 494.275274
min=  -1200.0




In [21]:
data_forct['Percentage_change']= ((30-data_forct['Stock_Qty'])/30)*100

In [34]:
data_forct[data_forct['Stocks']== 'Onhand'].max()


Date                       2022-11-27 00:00:00
Location               S SAN FRANCISCO BUS CTR
Item                               FILLET QQQQ
OH Quantity                            1019.95
On Order Quantity                       761.65
In Transit Quantity                     398.79
NSI Quantity                                 0
Quantity                                760.72
Amount                                 1002.36
Forecasted_Qty                          778.66
Stock_Qty                                 30.0
Stocks                                  Onhand
Stock_flag                                   3
Percentage_change                    89.066667
dtype: object

Onhand
max_change=  89.066667
min= 0.0

Overstock
max= -3.333333
min= -1153.333333

Understock:
max= 1460.0
min=  100.0

### Forecasted data

In [134]:
data_forct

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O        84.64   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O        65.47   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O        81.99   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O        82.37   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O        72.33   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        55.94   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        81.92   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        38.68   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        76.53   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        68.86   

     On Order Quantity  In Transit Quantity  NSI Quantity  Quantity  Amount  \
0                99.41                37.29             0     55.80  334.24   
1                71.47                11.99             0     41.94  251.21   
2                92.42                44.47             0     65.50  392.34   
3                98.76                42.11             0     27.73  166.11   
4                82.90                 4.71             0     19.68  117.88   
..                 ...                  ...           ...       ...     ...   
429              45.11                45.54             0    113.51  146.43   
430              47.40                68.25             0     97.40  125.65   
431              26.79                20.34             0     93.24  120.28   
432              76.63                64.16             0     94.38  121.75   
433              66.40                52.27             0     88.66  114.37   

     Forecasted_Qty  Stock_Qty     Stocks  Stock_flag  
0             46.71      24.00     Onhand           3  
1             54.25      48.00  Overstock           1  
2             49.95      16.00     Onhand           3  
3             38.10      12.00     Onhand           3  
4             43.47      49.00  Overstock           1  
..              ...        ...        ...         ...  
429          115.52      59.00  Overstock           1  
430          118.61      15.84     Onhand           3  
431          132.86     101.00  Overstock           1  
432           87.94      24.00     Onhand           3  
433           82.54      28.00     Onhand           3  

[434 rows x 13 columns]

In [145]:
data_forct["stock_percentage"] = (((data_forct["Forecasted_Qty"]-data_forct["Quantity"])/data_forct["Quantity"])*100).round(2)

In [146]:
data_forct

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O        84.64   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O        65.47   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O        81.99   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O        82.37   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O        72.33   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        55.94   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        81.92   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        38.68   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        76.53   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        68.86   

     On Order Quantity  In Transit Quantity  NSI Quantity  Quantity  Amount  \
0                99.41                37.29             0     55.80  334.24   
1                71.47                11.99             0     41.94  251.21   
2                92.42                44.47             0     65.50  392.34   
3                98.76                42.11             0     27.73  166.11   
4                82.90                 4.71             0     19.68  117.88   
..                 ...                  ...           ...       ...     ...   
429              45.11                45.54             0    113.51  146.43   
430              47.40                68.25             0     97.40  125.65   
431              26.79                20.34             0     93.24  120.28   
432              76.63                64.16             0     94.38  121.75   
433              66.40                52.27             0     88.66  114.37   

     Forecasted_Qty  Stock_Qty     Stocks  Stock_flag  stock_percentage  
0             46.71      24.00     Onhand           3            -16.29  
1             54.25      48.00  Overstock           1             29.35  
2             49.95      16.00     Onhand           3            -23.74  
3             38.10      12.00     Onhand           3             37.40  
4             43.47      49.00  Overstock           1            120.88  
..              ...        ...        ...         ...               ...  
429          115.52      59.00  Overstock           1              1.77  
430          118.61      15.84     Onhand           3             21.78  
431          132.86     101.00  Overstock           1             42.49  
432           87.94      24.00     Onhand           3             -6.82  
433           82.54      28.00     Onhand           3             -6.90  

[434 rows x 14 columns]

In [144]:
data_forct[data_forct['Stocks']== 'Understock'].min()

Date                   2022-05-01 00:00:00
Location                           FREMONT
Item                        BEEF OOOO - GT
OH Quantity                          20.46
On Order Quantity                      0.0
In Transit Quantity                    0.0
NSI Quantity                             0
Quantity                             17.81
Amount                               106.7
Forecasted_Qty                       16.43
Stock_Qty                           -408.0
Stocks                          Understock
Stock_flag                               2
stock_percentage                -39.653816
dtype: object

In [ ]:
Onhand
max_change=  81.61
min= -31.02

Overstock
max= 120.884146
min= -36.03716

Understock:
max= 60.613016
min=  -39.653816

In [147]:
data_forct= data_forct.drop(columns= ['Stocks', 'Stock_flag'], axis=0)

In [148]:
data_forct

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O        84.64   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O        65.47   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O        81.99   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O        82.37   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O        72.33   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        55.94   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        81.92   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        38.68   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        76.53   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        68.86   

     On Order Quantity  In Transit Quantity  NSI Quantity  Quantity  Amount  \
0                99.41                37.29             0     55.80  334.24   
1                71.47                11.99             0     41.94  251.21   
2                92.42                44.47             0     65.50  392.34   
3                98.76                42.11             0     27.73  166.11   
4                82.90                 4.71             0     19.68  117.88   
..                 ...                  ...           ...       ...     ...   
429              45.11                45.54             0    113.51  146.43   
430              47.40                68.25             0     97.40  125.65   
431              26.79                20.34             0     93.24  120.28   
432              76.63                64.16             0     94.38  121.75   
433              66.40                52.27             0     88.66  114.37   

     Forecasted_Qty  Stock_Qty  stock_percentage  
0             46.71      24.00            -16.29  
1             54.25      48.00             29.35  
2             49.95      16.00            -23.74  
3             38.10      12.00             37.40  
4             43.47      49.00            120.88  
..              ...        ...               ...  
429          115.52      59.00              1.77  
430          118.61      15.84             21.78  
431          132.86     101.00             42.49  
432           87.94      24.00             -6.82  
433           82.54      28.00             -6.90  

[434 rows x 12 columns]

In [168]:
conditions = [
    (data_forct['stock_percentage'] <= 0),
    (data_forct['stock_percentage'] > 0) & (data_forct['stock_percentage'] <= 25),
    (data_forct['stock_percentage'] > 25)]

In [169]:
values = ['Understock', 'Onhand', 'Overstock']
data_forct["stock"]=np.select(conditions, values)

In [170]:
len(data_forct[data_forct['stock']== 'Onhand'])

196

In [172]:
# Understock= 200 
# Onhand= 196
# Overstock= 84

In [175]:
def stock_flag(row):
    if row["stock"]== 'Overstock':
        return 1
    elif row["stock"]== 'Understock':
        return 2
    if row["stock"]== 'Onhand':
        return 3

In [176]:
data_forct["stock_flag"] = data_forct.apply(stock_flag, axis=1)

In [177]:
data_forct

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O        84.64   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O        65.47   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O        81.99   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O        82.37   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O        72.33   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        55.94   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        81.92   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        38.68   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        76.53   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        68.86   

     On Order Quantity  In Transit Quantity  NSI Quantity  Quantity  Amount  \
0                99.41                37.29             0     55.80  334.24   
1                71.47                11.99             0     41.94  251.21   
2                92.42                44.47             0     65.50  392.34   
3                98.76                42.11             0     27.73  166.11   
4                82.90                 4.71             0     19.68  117.88   
..                 ...                  ...           ...       ...     ...   
429              45.11                45.54             0    113.51  146.43   
430              47.40                68.25             0     97.40  125.65   
431              26.79                20.34             0     93.24  120.28   
432              76.63                64.16             0     94.38  121.75   
433              66.40                52.27             0     88.66  114.37   

     Forecasted_Qty  Stock_Qty  stock_percentage       stock  stock_flag  
0             46.71      24.00            -16.29  Understock           2  
1             54.25      48.00             29.35   Overstock           1  
2             49.95      16.00            -23.74  Understock           2  
3             38.10      12.00             37.40   Overstock           1  
4             43.47      49.00            120.88   Overstock           1  
..              ...        ...               ...         ...         ...  
429          115.52      59.00              1.77      Onhand           3  
430          118.61      15.84             21.78      Onhand           3  
431          132.86     101.00             42.49   Overstock           1  
432           87.94      24.00             -6.82  Understock           2  
433           82.54      28.00             -6.90  Understock           2  

[434 rows x 14 columns]

In [178]:
data_forct= data_forct.rename(columns= {'Quantity': 'Actual_Qty'})

In [180]:
data_forct

Date                 Location               Item  OH Quantity  \
0   2022-05-01                  FREMONT  CHICKEN XXXXX - O        84.64   
1   2022-05-08                  FREMONT  CHICKEN XXXXX - O        65.47   
2   2022-05-15                  FREMONT  CHICKEN XXXXX - O        81.99   
3   2022-05-22                  FREMONT  CHICKEN XXXXX - O        82.37   
4   2022-05-29                  FREMONT  CHICKEN XXXXX - O        72.33   
..         ...                      ...                ...          ...   
429 2022-10-30  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        55.94   
430 2022-11-06  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        81.92   
431 2022-11-13  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        38.68   
432 2022-11-20  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        76.53   
433 2022-11-27  S SAN FRANCISCO BUS CTR    DRUMSTICKS NNNN        68.86   

     On Order Quantity  In Transit Quantity  NSI Quantity  Actual_Qty  Amount  \
0                99.41                37.29             0       55.80  334.24   
1                71.47                11.99             0       41.94  251.21   
2                92.42                44.47             0       65.50  392.34   
3                98.76                42.11             0       27.73  166.11   
4                82.90                 4.71             0       19.68  117.88   
..                 ...                  ...           ...         ...     ...   
429              45.11                45.54             0      113.51  146.43   
430              47.40                68.25             0       97.40  125.65   
431              26.79                20.34             0       93.24  120.28   
432              76.63                64.16             0       94.38  121.75   
433              66.40                52.27             0       88.66  114.37   

     Forecasted_Qty  Stock_Qty  stock_percentage       stock  stock_flag  
0             46.71      24.00            -16.29  Understock           2  
1             54.25      48.00             29.35   Overstock           1  
2             49.95      16.00            -23.74  Understock           2  
3             38.10      12.00             37.40   Overstock           1  
4             43.47      49.00            120.88   Overstock           1  
..              ...        ...               ...         ...         ...  
429          115.52      59.00              1.77      Onhand           3  
430          118.61      15.84             21.78      Onhand           3  
431          132.86     101.00             42.49   Overstock           1  
432           87.94      24.00             -6.82  Understock           2  
433           82.54      28.00             -6.90  Understock           2  

[434 rows x 14 columns]

In [182]:
# data_forct.to_excel('forect_with_stock%22.xlsx', index=False)